File: python/editor.py
import sys
import pygame

from scripts.utils import load_images
from scripts.tilemap import Tilemap
from typing import Tuple

RENDER_SCALE = 2.0

DECOR_IMGS_PATH = 'tiles/decor'
GRASS_IMGS_PATH = 'tiles/grass'
LARGE_DECOR_IMGS_PATH = 'tiles/large_decor'
SPAWNERS_IMGS_PATH = 'tiles/spawners'
STONE_IMGS_PATH = 'tiles/stone'

class Editor:
  def __init__(self) -> None:
    pygame.init()
    pygame.display.set_caption("Editor")

    self.screen = pygame.display.set_mode((640, 480))
    self.display = pygame.Surface((320, 240))
    self.clock = pygame.time.Clock()

    self.assets = {
      'decor': load_images(DECOR_IMGS_PATH),
      'grass': load_images(GRASS_IMGS_PATH),
      'large_decor': load_images(LARGE_DECOR_IMGS_PATH),
      'stone': load_images(STONE_IMGS_PATH),
    }

    self.movement = [False, False, False, False]

    self.tilemap = Tilemap(self)

    try:
      self.tilemap.load('map.json')
    except FileNotFoundError:
      pass

    self.tile_list = list(self.assets)
    self.tile_group = 0
    self.tile_variant = 0

    self.clicking = False
    self.right_clicking = False
    self.shift = False
    self.ongrid = True

    self.scroll = [0, 0]

  def run(self):
    while True:
      self.display.fill((0, 0, 0))

      self.scroll[0] += (self.movement[1] - self.movement[0]) * 3
      self.scroll[1] += (self.movement[3] - self.movement[2]) * 3
      render_scroll = int(self.scroll[0]), int(self.scroll[1])
      self.tilemap.render(self.display, offset=render_scroll)

      current_tile_image = self.assets[self.tile_list[self.tile_group]][self.tile_variant].copy()
      current_tile_image.set_alpha(150)
      self.display.blit(current_tile_image, (5, 5))

      mouse_pos = pygame.mouse.get_pos()
      mouse_pos = (mouse_pos[0] / RENDER_SCALE, mouse_pos[1] / RENDER_SCALE)
      tile_pos = (int(mouse_pos[0] + self.scroll[0]) // self.tilemap.tile_size, int(mouse_pos[1] + self.scroll[1]) // self.tilemap.tile_size)

      if self.ongrid:
        self.display.blit(current_tile_image, (tile_pos[0] * self.tilemap.tile_size - self.scroll[0], tile_pos[1] * self.tilemap.tile_size - self.scroll[1]))
      else:
        self.display.blit(current_tile_image, mouse_pos)

      if self.clicking and self.ongrid:
        self.tilemap.tilemap[str(tile_pos[0]) + ';' + str(tile_pos[1])] = {
          'type': self.tile_list[self.tile_group],
          'variant': self.tile_variant,
          'pos': tile_pos
        }

      if self.right_clicking:
        tile_location = str(tile_pos[0]) + ';' + str(tile_pos[1])
        if tile_location in self.tilemap.tilemap:
          del self.tilemap.tilemap[tile_location]
        for tile in self.tilemap.offgrid_tiles.copy():
          tile_img = self.assets[tile['type']][tile['variant']]
          tile_rect = pygame.Rect(tile['pos'][0] - self.scroll[0], tile['pos'][1] - self.scroll[1], tile_img.get_width(), tile_img.get_height())
          if tile_rect.collidepoint(mouse_pos):
            self.tilemap.offgrid_tiles.remove(tile)

      for event in pygame.event.get():
        if event.type == pygame.QUIT:
          self.handle_quit()

        if event.type == pygame.MOUSEBUTTONDOWN:
          self.handle_mouse_down(event.button, mouse_pos=mouse_pos, render_scroll=render_scroll)

        if event.type == pygame.MOUSEBUTTONUP:
          self.handle_mouse_up(event.button)

        if event.type == pygame.KEYDOWN:
          self.handle_key_down(event.key)

        if event.type == pygame.KEYUP:
          self.handle_key_up(event.key)

      self.screen.blit(
        pygame.transform.scale(self.display, self.screen.get_size()),
        (0, 0)
      )
      pygame.display.update()
      self.clock.tick(60)

  def handle_mouse_up(self, button: int) -> None:
    if button == 1:
      self.clicking = False

    if button == 3:
      self.right_clicking = False

  def handle_mouse_down(self, button: int, mouse_pos: Tuple[int, int]=None, render_scroll: Tuple[int, int]=None) -> None:
    if button == 1:
      self.clicking = True
      if not self.ongrid:
        self.tilemap.offgrid_tiles.append({
          'type': self.tile_list[self.tile_group],
          'variant': self.tile_variant,
          'pos': (mouse_pos[0] + render_scroll[0], mouse_pos[1] + render_scroll[1])
        })

    if button == 3:
      self.right_clicking = True

    if self.shift:
      if button == 4:
        self.tile_variant = (self.tile_variant + 1) % len(self.assets[self.tile_list[self.tile_group]]) 

      if button == 5:
        self.tile_variant = (self.tile_variant - 1) % len(self.assets[self.tile_list[self.tile_group]])
    else:
      if button == 4:
        self.tile_group = (self.tile_group + 1) % len(self.tile_list)
        self.tile_variant = 0

      if button == 5:
        self.tile_group = (self.tile_group - 1) % len(self.tile_list)
        self.tile_variant = 0

  def handle_key_down(self, key: int) -> None:
    if key == pygame.K_a:
      self.movement[0] = True

    if key == pygame.K_d:
      self.movement[1] = True

    if key == pygame.K_w:
      self.movement[2] = True

    if key == pygame.K_s:
      self.movement[3] = True

    if key == pygame.K_LSHIFT:
      self.shift = True

    if key == pygame.K_o:
      self.tilemap.save('map.json')

    if key == pygame.K_t:
      self.tilemap.auto_tile()

    if key == pygame.K_g:
      self.ongrid = not self.ongrid

  def handle_key_up(self, key: int) -> None:
    if key == pygame.K_a:
      self.movement[0] = False

    if key == pygame.K_d:
      self.movement[1] = False

    if key == pygame.K_w:
      self.movement[2] = False

    if key == pygame.K_s:
      self.movement[3] = False

    if key == pygame.K_LSHIFT:
      self.shift = False

  def handle_quit(self):
    pygame.quit()
    sys.exit()

Editor().run()

File: golang/editor/editor.go
package main

import (
	"github.com/hajimehoshi/ebiten/v2"
	"github.com/hajimehoshi/ebiten/v2/inpututil"
	"github.com/yuricorredor/platformer/assets"
	"github.com/yuricorredor/platformer/rects"
	"github.com/yuricorredor/platformer/tilemap"
	"github.com/yuricorredor/platformer/types"
)

var (
	RENDER_SCALE   = 2
	MOVEMENT_SPEED = 3
	PATH           = "assets/maps/new_map.json"
)

type Editor struct {
	scrollX       int
	scrollY       int
	screenWidth   int
	screenHeight  int
	clicking      bool
	rightClicking bool
	onGrid        bool
	tileList      []string
	tileGroup     int
	tileVariant   int
	position      types.Vector
}

func (e *Editor) Update() error {
	e.HandleCursor()
	e.HandleInputs()

	return nil
}

func (e *Editor) Draw(screen *ebiten.Image) {
	tilemap.TileMap.Draw(screen, e.scrollX, e.scrollY)

	e.DrawLayout(screen)
	e.DrawCurrentTile(screen)
}

func (e *Editor) CurrentTileImage() *ebiten.Image {
	return assets.Assets.Images[e.tileList[e.tileGroup]][e.tileVariant]
}

func (e *Editor) DrawLayout(screen *ebiten.Image) {
	currentTileImage := e.CurrentTileImage()

	options := &ebiten.DrawImageOptions{}
	options.ColorScale.ScaleAlpha(0.65)
	options.GeoM.Translate(5, 5)
	screen.DrawImage(currentTileImage, options)
}

func (e *Editor) DrawCurrentTile(screen *ebiten.Image) {
	currentTileImage := e.CurrentTileImage()
	options := &ebiten.DrawImageOptions{}
	options.ColorScale.ScaleAlpha(0.65)

	if e.onGrid {
		options.GeoM.Translate(float64((int(e.position.X)+e.scrollX)/tilemap.TileMap.TileSize)*float64(tilemap.TileMap.TileSize)-float64(e.scrollX), float64((int(e.position.Y)+e.scrollY)/tilemap.TileMap.TileSize)*float64(tilemap.TileMap.TileSize)-float64(e.scrollY))
	} else {
		options.GeoM.Translate(e.position.X-float64(currentTileImage.Bounds().Dx()/2), e.position.Y-float64(currentTileImage.Bounds().Dy()/2))
	}

	screen.DrawImage(currentTileImage, options)
}

func (e *Editor) HandleCursor() {
	mouseX, mouseY := ebiten.CursorPosition()
	isCursorInsideGameScreen := mouseX >= 0 && mouseX < e.screenWidth && mouseY >= 0 && mouseY < e.screenHeight
	if isCursorInsideGameScreen {
		e.position = types.Vector{X: float64(mouseX), Y: float64(mouseY)}
	}
}

func (e *Editor) HandleInputs() {
	if inpututil.IsKeyJustPressed(ebiten.KeyG) {
		e.onGrid = !e.onGrid
	}
	if inpututil.IsKeyJustPressed(ebiten.KeyO) {
		tilemap.TileMap.Save(PATH)
	}

	if ebiten.IsKeyPressed(ebiten.KeyW) {
		e.scrollY -= MOVEMENT_SPEED
	}
	if ebiten.IsKeyPressed(ebiten.KeyS) {
		e.scrollY += MOVEMENT_SPEED
	}
	if ebiten.IsKeyPressed(ebiten.KeyA) {
		e.scrollX -= MOVEMENT_SPEED
	}
	if ebiten.IsKeyPressed(ebiten.KeyD) {
		e.scrollX += MOVEMENT_SPEED
	}

	if ebiten.IsMouseButtonPressed(ebiten.MouseButtonLeft) {
		e.clicking = true
	} else {
		e.clicking = false
	}
	if ebiten.IsMouseButtonPressed(ebiten.MouseButtonRight) {
		e.rightClicking = true
	} else {
		e.rightClicking = false
	}
	_, yoff := ebiten.Wheel()

	if ebiten.IsKeyPressed(ebiten.KeyShift) {
		if yoff > 0 {
			e.tileVariant++
			if e.tileVariant >= len(assets.Assets.Images[e.tileList[e.tileGroup]]) {
				e.tileVariant = 0
			}
		}
		if yoff < 0 {
			e.tileVariant--
			if e.tileVariant < 0 {
				e.tileVariant = len(assets.Assets.Images[e.tileList[e.tileGroup]]) - 1
			}
		}
	} else {
		if yoff > 0 {
			e.tileGroup++
			if e.tileGroup >= len(e.tileList) {
				e.tileGroup = 0
			}
			e.tileVariant = 0
		}
		if yoff < 0 {
			e.tileGroup--
			if e.tileGroup < 0 {
				e.tileGroup = len(e.tileList) - 1
			}
			e.tileVariant = 0
		}
	}

	if ebiten.IsMouseButtonPressed(ebiten.MouseButtonRight) {
		e.RemoveTile()
		e.RemoveOffGridTile()
	}

	if inpututil.IsMouseButtonJustPressed(ebiten.MouseButtonLeft) && !e.onGrid {
		e.AddOffgridTile()
	}

	if e.clicking && e.onGrid {
		e.AddTile()
	}
}

func (e *Editor) RemoveTile() {
	tilemap.TileMap.RemoveTile(types.Vector{X: float64((int(e.position.X) + e.scrollX) / tilemap.TileMap.TileSize), Y: float64((int(e.position.Y) + e.scrollY) / tilemap.TileMap.TileSize)})
}

func (e *Editor) RemoveOffGridTile() {
	// for tile in self.tilemap.offgrid_tiles.copy():
	//         tile_img = self.assets[tile['type']][tile['variant']]
	//         tile_rect = pygame.Rect(tile['pos'][0] - self.scroll[0], tile['pos'][1] - self.scroll[1], tile_img.get_width(), tile_img.get_height())
	//         if tile_rect.collidepoint(mouse_pos):
	//           self.tilemap.offgrid_tiles.remove(tile)
	for _, tile := range tilemap.TileMap.OffGridTiles {
		tile_image := assets.Assets.Images[tile.Type][tile.Variant]
		tileRect := rects.Rect{
			X:      tile.Position.X*float64(tilemap.TileMap.TileSize) - float64(e.scrollX),
			Y:      tile.Position.Y*float64(tilemap.TileMap.TileSize) - float64(e.scrollY),
			Width:  float64(tile_image.Bounds().Dx()),
			Height: float64(tile_image.Bounds().Dy()),
		}

		if tileRect.Contains(e.position) {
			tilemap.TileMap.RemoveOffGridTile(tile)
		}
	}
}

func (e *Editor) AddTile() {
	tile := tilemap.Tile{
		Position: types.Vector{X: float64((int(e.position.X) + e.scrollX) / tilemap.TileMap.TileSize), Y: float64((int(e.position.Y) + e.scrollY) / tilemap.TileMap.TileSize)},
		Variant:  e.tileVariant,
		Type:     e.tileList[e.tileGroup],
	}

	tilemap.TileMap.SetTile(tile)
}

func (e *Editor) AddOffgridTile() {
	tile := tilemap.Tile{
		Position: types.Vector{X: (e.position.X + float64(e.scrollX) - float64(e.CurrentTileImage().Bounds().Dx()/2)) / float64(tilemap.TileMap.TileSize), Y: (e.position.Y + float64(e.scrollY) - float64(e.CurrentTileImage().Bounds().Dy()/2)) / float64(tilemap.TileMap.TileSize)},
		Variant:  e.tileVariant,
		Type:     e.tileList[e.tileGroup],
	}

	tilemap.TileMap.SetOffGridTile(tile)
}

func (e *Editor) Layout(outsideWidth, outsideHeight int) (screenWidth, screenHeight int) {
	e.screenWidth = outsideWidth / RENDER_SCALE
	e.screenHeight = outsideHeight / RENDER_SCALE
	return e.screenWidth, e.screenHeight
}

func NewEditor() *Editor {
	return &Editor{
		onGrid: true,
	}
}

func main() {
	editor := NewEditor()

	tilemap.TileMap.Load(PATH)

	editor.tileList = []string{"grass", "stone", "decor", "large_decor"}

	ebiten.SetCursorMode(ebiten.CursorModeHidden)
	ebiten.SetWindowSize(640, 480)
	ebiten.SetWindowTitle("Platformer Editor")

	if err := ebiten.RunGame(editor); err != nil {
		panic(err)
	}
}
